### ISSUE TITLE
About the use of ensure_xcm()

### ISSUE THREAD

**Issue by hqwangningbo** [paritytech/polkadot#7493]
I would like to ask whether ensure_xcm() can be used normally at present. 
I have not seen an example of use yet.
Because I saw the example of ensure_response() call, I tried to use ensure_xcm() like its code.
Send Xcm Transact instruction from other chains to this chain. 
But the result was BadOrigin.

Thanks for answering🙏

**Comment by bkchr** [comment]
CC @KiChjang

**Comment by KiChjang** [comment]
In order to make use of `ensure_xcm`, let's first take a look at the code:

```rust
pub fn ensure_xcm<OuterOrigin>(o: OuterOrigin) -> Result<MultiLocation, BadOrigin>
// ...
{
	match o.into() {
		Ok(Origin::Xcm(location)) => Ok(location),
		// ...
	}
}
```

Here you can see that it's trying to convert the `OuterOrigin` (we call it `RuntimeOrigin` nowadays) into an `Origin::Xcm` variant. This means that the `MultiLocation` origin must have been converted to a FRAME origin before by an origin converter. Currently, there is a single origin converter that does this, and it's called `XcmPassthrough`, located in the XCM pallet.

In short, `ensure_xcm` only works if you have specifically configured your `OriginConverter` to be `XcmPassthrough`, otherwise you will never be able to get an `Origin::Xcm` to match against.

**Comment by hqwangningbo** [comment]
@KiChjang Ok, so that's the case, thanks for the answer.

I would like to ask, for example, if I want to write a pallet with a test_call method in it, and another chain calls it via DescendOrigin+Trancact instruction, what operation can I use to get the para_id, caller, and other useful information of the calling chain in the test_call method ?

**Comment by KiChjang** [comment]
The most direct way of doing so is to include them as parameters to `test_method`. Outside of that, `XcmPassthrough` has the `MultiLocation` of the sender, so you can quickly check and see which parachain it came from by matching agains the `Parachain` junction.
